rq/worker.py

```diff
@@ -568,86 +722,146 @@ def prepare_job_execution(self, job):
-                    rv = job.perform()
+                pipeline.execute()
+            except Exception:
+                # Ensure that custom exception handlers are called
+                # even if Redis is down
+                pass
 
-                # Pickle the result in the same try-except block since we need
-                # to use the same exc handling when pickling fails
-                job._result = rv
+    def handle_job_success(self, job, queue, started_job_registry):
 
-                self.set_current_job_id(None, pipeline=pipeline)
+        with self.connection._pipeline() as pipeline:
+            while True:
+                try:
+                    # if dependencies are inserted after enqueue_dependents
+                    # a WatchError is thrown by execute()
+                    pipeline.watch(job.dependents_key)
+                    # enqueue_dependents calls multi() on the pipeline!
+                    queue.enqueue_dependents(job, pipeline=pipeline)
+
+                    self.set_current_job_id(None, pipeline=pipeline)
+                    self.increment_successful_job_count(pipeline=pipeline)
+                    self.increment_total_working_time(
+                        job.ended_at - job.started_at, pipeline
+                    )
+
+                    result_ttl = job.get_result_ttl(self.default_result_ttl)
+                    if result_ttl != 0:
+                        job.set_status(JobStatus.FINISHED, pipeline=pipeline)
+                        # Don't clobber the user's meta dictionary!
+                        job.save(pipeline=pipeline, include_meta=False)
+
+                        finished_job_registry = FinishedJobRegistry(job.origin,
+                                                                    self.connection,
+                                                                    job_class=self.job_class)
+                        finished_job_registry.add(job, result_ttl, pipeline)
+
+                    job.cleanup(result_ttl, pipeline=pipeline,
+                                remove_from_queue=False)
+                    started_job_registry.remove(job, pipeline=pipeline)
 
-                result_ttl = job.get_result_ttl(self.default_result_ttl)
-                if result_ttl != 0:
-                    job.ended_at = utcnow()
-                    job._status = JobStatus.FINISHED
-                    job.save(pipeline=pipeline)
+                    pipeline.execute()
 
-                    finished_job_registry = FinishedJobRegistry(job.origin, self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                    # nuglab custom finish handler
+                    logger.info('Call %d handlers on successful finish of job %s', len(self.finish_handlers), job.id)
+                    for f in self.finish_handlers:
+                        f(job, self.connection)
```

Comment:
in the original commit (https://github.com/rq/rq/commit/4961f38c3fa84435a574e9bcc6436a67a726808a) this was called in perform_job(). But the latest rq has its own handle_job_success() which seems more appropriate for adding custom finisher logic.